### PR TITLE
[Docs] Adding Store API Docs

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45543 [Docs] Adding Store API Docs**

This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

See example render below:
Screen Shot 2020-09-29 at 11 42 33 PM

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
...
+
+Arguments:
+    key (str): The key to be deleted from the store
+
+Returns:
+    `true` if ``key`` was deleted, otherwise `false`.
+
+Example::
+    >>> import torch.distributed as dist
+    >>> # Using TCPStore as an example, other store types can also be used
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/init.cpp:394]: nit: Other store types won't work with this example due to delete_key exceptions right?

Comment (osalpekar) [reply torch/csrc/distributed/c10d/init.cpp:394]: Right, good catch

File: docs/source/distributed.rst [hunk 2]
```diff
@@ -270,6 +270,30 @@ The machine with rank 0 will be used to set up all connections.
...
+
+The distributed package comes with a distributed key-value store, which can be
+used to share information between processes in the group as well as to
+initialize the distributed pacakge in
+:func:`torch.distributed.init_process_group` (by explicitly creating the store
+as an alternative to specifying ``init_method``.) There are 3 choices for
+Key-Value Stores: :class:`~torch.distributed.TCPStore`,
+:class:`~torch.distributed.FileStore`, and :class:`~torch.distributed.HashStore`.
+
+.. autoclass:: TCPStore
```
Comment (pritamdamania87) [inline docs/source/distributed.rst:284]: Should we also add Store here? Similar to torch.optim where there is documentation about the base class.

File: torch/csrc/distributed/c10d/init.cpp [hunk 3]
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
...
               "add",
               &::c10d::Store::add,
-              py::call_guard<py::gil_scoped_release>())
+              py::call_guard<py::gil_scoped_release>(),
+              R"(
+add(key: str, amount: int) -> None
+
+The first call to add for a given ``key`` creates a counter associated
+with ``key`` in the store, initialized to ``amount``. Subsequent calls to add
+with the same ``key`` increment the counter by the specified ``amount``.
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:358]: What is the behavior if the key is already present via `set` call? We should document that as well.

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
...
           .def(
               "num_keys",
               &::c10d::Store::getNumKeys,
-              py::call_guard<py::gil_scoped_release>())
+              py::call_guard<py::gil_scoped_release>(),
+              R"(
+num_keys() -> int
+
+Returns the number of keys set in the store. Note that this number will typically
+be one greater than the number of keys added by :meth:`~torch.distributed.store.get`
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:410]: This should be `set` and not `get`?

File: torch/csrc/distributed/c10d/init.cpp [hunk 5]
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
...
+    >>> store2.get("first_key")
+
+      )")
       .def(py::init<const std::string&, int>());
 
 #ifndef _WIN32
-  shared_ptr_class_<::c10d::HashStore>(module, "HashStore", store)
+  shared_ptr_class_<::c10d::HashStore>(module, "HashStore", store,
+      R"(
+A thread-safe store implementation based on an underlying hashmap (C++ std::unordered_map).
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:520]: Don't add implementation details like c++ unordered_map. Also, mention that this store can only be used within the same process and not across processes.

File: torch/csrc/distributed/c10d/init.cpp (hunk 3 above)
Comment (osalpekar) [reply torch/csrc/distributed/c10d/init.cpp:358]: It seems to throw an exception - will make a note about it in the docs.

File: torch/csrc/distributed/c10d/init.cpp [hunk 6]
```diff
@@ -284,7 +284,11 @@ They are used in specifying strategies for reduction collectives, e.g.,
 
   auto store =
       py::class_<::c10d::Store, std::shared_ptr<::c10d::Store>, PythonStore>(
-          module, "Store")
+          module, "Store",
+          R"(
+Base class for the 3 store implementations: (:class:`~torch.distributed.TCPStore`,
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:289]: Probably just mention this is the base class for all store implementations. There could be other user implementations on top of this apart from these 3.

**Review by pritamdamania87 — APPROVED** [review]

**Pull request closed without merging**